docs/debug.md

```diff
@@ -0,0 +1,298 @@
+(gdb)
+```
+
+## 1.5. Use graphical frontends
+### 1.5.1 ddd
+With the `PATH` exported to the `arm-none-eabi-gdb` binary and the `optee`
+helper function defined as above in the `.gdbinit` file, you invoke ddd by
+typing:
+
+```
+ddd --debugger arm-none-eabi-gdb
+```
+
+Then in the lower pane (which is the gdb command window), just simply type
+`optee` and ddd will connect to the remote and load `tee.elf`, just as described
+above for the command line version.
+
+### 1.5.2 GNU Visual Debugger ([gvd](http://gnu.gds.tuwien.ac.at/directory/gnuVisualDebugger.html))
+This is a rather old frontend for gdb and share a lot of similarities with ddd,
+however it seems like it's more stable compared to ddd. To run it, you simply
+need to tell the path to the `arm-none-eabi-gdb` binary:
+
+```
+gvd --debugger $HOME/devel/toolchains/gcc-linaro-arm-none-eabi-4.9-2014.09_linux/bin/arm-none-eabi-gdb
+```
+
+Similarly to ddd, just simply run `optee` in the lower gdb command pane in gvd.
+
+### 1.5.3 Affinic Debugger / ADG
+[Affinic Debugger] seems to be the most stable and graphical front end. It's not
+free (at this moment it costs roughly $50 USD). If you can afford it and prefer
+graphical frontends we highly recommends this tool.
+
+# 2. Ftrace
+Ftrace is useful set of tools for debugging both kernel and to some extent user
+space. Ftrace is really useful if you want to learn how some piece of code
+interact with other parts of the system. It's nothing special you have to do to
+make use of ftrace for OP-TEE. But for a reference we list a couple of commands
+and scenarios that could be handy to have ready to be copy/pasted.
+
+## 2.1 Enable ftrace in menuconfig
+First you will need to enable ftrace in the kernel. Depending on which version
+you are using it might look a bit different compared to what is shown below
+(here we were using `4.1.0-rc4`)
+
+```
+make ARCH=arm menuconfig
+    # Go into "Kernel hacking"
+    General setup  --->
+    ...
+    Kernel hacking  --->
+
+    # Enable and go into Tracers
+    ...
+    [*] Tracers  --->
+
+    # Below is a good set of features (*) to enable
+    --- Tracers
+    -*-   Kernel Function Tracer
+    [*]     Kernel Function Graph Tracer
+    [ ]   Interrupts-off Latency Tracer
+    [ ]   Scheduling Latency Tracer
+    [*]   Trace syscalls
+    [ ]   Create a snapshot trace buffer
+          Branch Profiling (No branch profiling)  --->
+    [*]   Trace max stack
+    [ ]   Support for tracing block IO actions
+    [ ]   Enable uprobes-based dynamic events
+    [*]   enable/disable function tracing dynamically
+    [*]   Kernel function profiler
+    [ ]   Perform a startup test on ftrace
+    [ ]   Add tracepoint that benchmarks tracepoints
+    < >   Ring buffer benchmark stress tester
+    [ ]   Ring buffer startup self test
+    [ ]   Show enum mappings for trace events
+```
+
+Then simply recompile the kernel.
+
+## 2.2 Use cases
+### 2.2.1 Filter OP-TEE functions
+```bash
+$ modprobe optee_armtz
```

Comment:
This should be replace with something more up-to-date, since we don't have really use this any longer. But I think I'll keep it for now and deal with that in a separate "cleanup ftrace" instructions patch.